### ISSUE TITLE
[BUG] `lower_bound` not working properly with list columns containing `null` entries

### ISSUE THREAD

**Issue by cindyyuanjiang** [rapidsai/cudf#12389]
**Describe the bug**
`lower_bound` produces incorrect results for list columns containing `null` entries.

**Steps/Code to reproduce bug**
Here is a reproducible test in cudf:
```
{
  using T = TypeParam;
  using lcw = cudf::test::lists_column_wrapper<T, int32_t>;
  using cudf::test::iterators::nulls_at;
  lcw col1{
    lcw{{0}, nulls_at({0})},      // 0
    lcw{-80},                     // 1
    lcw{-17},                     // 2
  };

  lcw col2{
    lcw{27},                      // 0
    lcw{{0}, nulls_at({0})},      // 1
    lcw{},                        // 2
  };

  lcw val1{
    lcw{87},
  };

  lcw val2{
    lcw{},
  };

  cudf::table_view input{{col1, col2}};
  cudf::table_view values{{val1, val2}};
  std::vector<cudf::order> column_order{cudf::order::ASCENDING, cudf::order::DESCENDING};
  std::vector<cudf::null_order> null_order_flags{cudf::null_order::BEFORE, cudf::null_order::BEFORE};
  
  auto expect = cudf::test::fixed_width_column_wrapper<cudf::size_type>{3};
  auto result = cudf::lower_bound(input, values, column_order, null_order_flags);
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expect, *result);
}
```

The test fails with output:
`first difference: lhs[0] = 3, rhs[0] = 1`

In other words, `([87], [])` should be in Partition 3, but instead was in Partition 1.

Note that if we remove the `null` entries, the test would pass, as shown below.
```
{
  using T = TypeParam;
  using lcw = cudf::test::lists_column_wrapper<T, int32_t>;
  using cudf::test::iterators::nulls_at;
  lcw col1{
    lcw{},                        // 0
    lcw{-80},                     // 1
    lcw{-17},                     // 2
  };

  lcw col2{
    lcw{27},                      // 0
    lcw{},                        // 1
    lcw{},                        // 2
  };

  lcw val1{
    lcw{87},
  };

  lcw val2{
    lcw{},
  };

  cudf::table_view input{{col1, col2}};
  cudf::table_view values{{val1, val2}};
  std::vector<cudf::order> column_order{cudf::order::ASCENDING, cudf::order::DESCENDING};
  std::vector<cudf::null_order> null_order_flags{cudf::null_order::BEFORE, cudf::null_order::BEFORE};
  
  auto expect = cudf::test::fixed_width_column_wrapper<cudf::size_type>{3};
  auto result = cudf::lower_bound(input, values, column_order, null_order_flags);
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expect, *result);
}
```

**Comment by ttnghia** [comment]
This may be related to https://github.com/rapidsai/cudf/issues/12298 since both share the same list comparator.

**Comment by GregoryKimball** [comment]
Is the root cause that the list lexicographic comparator does not respect the requested `null_order` value? Does `cudf::null_order::AFTER` work as expected?

**Comment by ttnghia** [comment]
@GregoryKimball Please have a look at the examples in https://github.com/rapidsai/cudf/issues/12298. There is a bug when comparing lists containing null. Currently, the lex comparator outputs `[[null]] < [[null, null]]` but (incorrectly) outputs `[[null, null]] < [[null], [-21827]]`.

**Comment by GregoryKimball** [comment]
I believe that this is solved by #12538, please feel free to reopen or submit an new issue if you encounter other issues with `lower_bound`.

**Comment by divyegala** [comment]
@GregoryKimball @cindyyuanjiang @ttnghia I will just mention that the results look right to me anyway, because the original example in this issue does not have any top-level NULLs. I suspect the expected behavior will be achieved with that change

**Comment by ttnghia** [comment]
This will be verified in https://github.com/NVIDIA/spark-rapids/pull/7233. If the issue still persists then we will reopen this.

Actually this should be closed only after confirmation from PR#7233. But it is fine to not reopen it for now.

**Comment by cindyyuanjiang** [comment]
I encountered a similar issue today.

Here is a reproducible test in cudf:
```
{
  using lcw = cudf::test::lists_column_wrapper<double>;
  using cudf::test::iterators::nulls_at;
  lcw col1{
    lcw{-3.45967821e+12}, // 0
    lcw{-3.6912186e-32},  // 1
    lcw{9.721175},        // 2
  };

  lcw val1{
    lcw{{0, 4.22671e+32}, nulls_at({0})},
  };

  cudf::table_view input{{col1}};
  cudf::table_view values{{val1}};
  std::vector<cudf::order> column_order{cudf::order::ASCENDING};
  std::vector<cudf::null_order> null_order_flags{cudf::null_order::BEFORE};
  
  auto expect = cudf::test::fixed_width_column_wrapper<cudf::size_type>{0};
  auto result = cudf::lower_bound(input, values, column_order, null_order_flags);
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expect, *result);
}
```

This fails with output: `first difference: lhs[0] = 0, rhs[0] = 3`
`[null, 4.22671e+32]` should be in Partition 0, but instead was in Partition 3.

**Comment by divyegala** [comment]
Thanks @cindyyuanjiang, I am reopening this issue. I can confirm that this is a `lower_bound` specific bug and not a list-lex bug because:
```
 {
    lcw col1{
      lcw{-3.45967821e+12}, // 0
      lcw{-3.6912186e-32},  // 1
      lcw{9.721175},        // 2
      lcw{{0, 4.22671e+32}, nulls_at({0})} // 3
    };

    auto result = cudf::sorted_order(cudf::table_view({col1}));
    cudf::test::print(*result);
 }

3,0,1,2
```

So it looks like the comparisons are fine as you can see the `needle` from your example gets sorted to position 0, and you would expect `lower_bound` to produce the same results

**Comment by bdice** [comment]
@divyegala I started looking into this a bit but didn't get very far. I think we should isolate some pieces. First, I would test the behavior of the `two_table_comparator` with lists containing nulls and ensure that it gives the expected results. We don't have many algorithms using the lexicographic `two_table_comparator` so `lower_bound` (and `upper_bound`) may be the only places where we'd see the problem. Let's open a draft PR with some of these known failing cases so that we can iterate together and see CI logs.

**Comment by divyegala** [comment]
My understanding of why this bug occurs is that since Dremel encoding is an independent, per-table process, the encoding for these examples end up looking like:
`haystack`
```
lcw col1{
  lcw{-3.45967821e+12}, // 0
  lcw{-3.6912186e-32},  // 1
  lcw{9.721175},        // 2
};

def_level = [1, 1, 1]
```
`needle`
``` 
lcw val1{
    lcw{{0, 4.22671e+32}, nulls_at({0})},
};
def_level = [1, 2]
```

This is happening because `nullability` plays an important role in the total number of definition levels, but because the `haystack` isn't `nullable`, the leaf in the `haystack` ends up at the same definition level as the `NULL` in the `needle`. Thus, the iteration continues and all the rows in the `haystack`, on account of being shorter length than the row in the `needle`, end up comparing as lexicographically `LESS` [here](https://github.com/rapidsai/cudf/blob/a1451e32b31a5b053ac113485ce17c22d0a28a75/cpp/include/cudf/table/experimental/row_operators.cuh#L483)

The fix for this is to normalize the Dremel encoding such that there's a preset mapping from elements to definition levels, regardless of `nullability` at any depth, such that:
```
0 - top level empty list
1 - top level null
2 - first level nested empty list
3 - first level nested null
...
n - leaf element
```

Finally, the `two_table_comparator` should have a constraint that both tables have the same nesting depth.

cc @bdice

**Comment by divyegala** [comment]
@cindyyuanjiang I have a PR up that fixes this https://github.com/rapidsai/cudf/pull/12726. Would you mind running integration tests with this PR and letting me know if the bug still persists or if you find other test cases that fail?
